service/s3/s3manager/batch_test.go

```diff
@@ -590,8 +592,141 @@ func TestBatchError(t *testing.T) {
 		t.Error("Expected error, but received nil")
 	}
 
-	if svc.index != len(objects) {
-		t.Errorf("Expected %d, but received %d", len(objects), svc.index)
+	if index != len(objects) {
+		t.Errorf("Expected %d, but received %d", len(objects), index)
+	}
+
+}
+
+type testAfterIter struct {
+	afterDelete   bool
+	afterDownload bool
+	afterUpload   bool
+	index         int
+}
+
+func (iter *testAfterIter) Next() bool {
+	next := (iter.index & 1) == 0
+	iter.index++
+	return next
+}
+
+func (iter *testAfterIter) Err() error {
+	return nil
+}
+
+func (iter *testAfterIter) DeleteObject() BatchDeleteObject {
+	return BatchDeleteObject{
+		Object: &s3.DeleteObjectInput{
+			Bucket: aws.String("foo"),
+			Key:    aws.String("foo"),
+		},
+		After: func() error {
+			iter.afterDelete = true
+			return nil
+		},
+	}
+}
+
+func (iter *testAfterIter) DownloadObject() BatchDownloadObject {
+	return BatchDownloadObject{
+		Object: &s3.GetObjectInput{
+			Bucket: aws.String("foo"),
+			Key:    aws.String("foo"),
+		},
+		Writer: aws.NewWriteAtBuffer([]byte{}),
+		After: func() error {
+			iter.afterDownload = true
+			return nil
+		},
 	}
+}
 
+func (iter *testAfterIter) UploadObject() BatchUploadObject {
+	return BatchUploadObject{
+		Object: &UploadInput{
+			Bucket: aws.String("foo"),
+			Key:    aws.String("foo"),
+			Body:   strings.NewReader("bar"),
+		},
+		After: func() error {
+			iter.afterUpload = true
+			return nil
+		},
+	}
+}
+
+func TestAfter(t *testing.T) {
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+	}))
+
+	sess := session.New(&aws.Config{
+		Endpoint:         &server.URL,
+		S3ForcePathStyle: aws.Bool(true),
+		Region:           aws.String("foo"),
+		Credentials:      credentials.NewStaticCredentials("AKID", "SECRET", "SESSION"),
+	})
+
+	index := 0
+	responses := []response{
+		{
+			&s3.PutObjectOutput{},
+			nil,
+		},
+		{
+			&s3.GetObjectOutput{},
+			nil,
+		},
+		{
+			&s3.DeleteObjectOutput{},
+			nil,
+		},
+	}
+
+	svc := &mockClient{
+		S3API: s3.New(sess),
+		Put: func() (*s3.PutObjectOutput, error) {
+			resp := responses[index]
+			index++
+			return resp.out.(*s3.PutObjectOutput), resp.err
+		},
+		Get: func() (*s3.GetObjectOutput, error) {
+			resp := responses[index]
+			index++
+			return resp.out.(*s3.GetObjectOutput), resp.err
+		},
+		List: func() (*s3.ListObjectsOutput, error) {
+			resp := responses[index]
+			index++
+			return resp.out.(*s3.ListObjectsOutput), resp.err
+		},
+	}
+	uploader := NewUploaderWithClient(svc)
+	downloader := NewDownloaderWithClient(svc)
+	deleter := NewBatchDeleteWithClient(svc)
+
+	iter := &testAfterIter{}
```

Comment:
should create a new iterator for each test case to prevent pollution. Sharing the iterator value can hide or introduce bugs in the test.